### PR TITLE
fix(behavior_velocity): occlusion spot offset

### PR DESCRIPTION
**Pull request by taikitanaka3**
## Related Issue(required)

resolve part of 1~3 #258

## Description(required)

1. fix offset for arc length 
2. ~make less false positive using lanelet info~ will be another PR
3. fix spline interpolation v

=
1. without path behind ego , interpolation can be linear at curve and will make worse result 
 image
2. it's necessary to remove vehicle waiting for traffic as occlusion spot
3. fix interpolation of velocity 

others
- calculate distance based on vehicle polygon not approximation.
- ~for the reason of changing slice size is lower computation cost a lot without lack of accuracy since occlusion spot is selected for each slice.~
- use nearest occlusion to path at each slice not "occlusion found at first".
- fix velocity of spline interpolation 

## Review Procedure(required)

1. velocity set around ego path is accurate in order
image

2. check if occlusion spot (s,d,v) is same as assumption
image

3. be sure private module / public module works at kashiwanoha
image

## Related PR(optional)
https://github.com/tier4/autoware.iv/pull/2283
https://github.com/autowarefoundation/autoware.universe/pull/249#issue-1101379992

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [ ] Read [commit-guidelines][commit-guidelines]
- [ ] Assign PR to reviewer

If you are adding new package following items are required:

- [ ] Documentation with description of the package is available
- [ ] A sample launch file and parameter file are available if the package contains executable nodes

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets

## CI Checks

- **Build and test for PR / build-and-test-pr**: Required to pass before the merge.
- **Build and test for PR / clang-tidy-pr**: NOT required to pass before the merge. It is up to the reviewer(s). Found false positives? See the [guidelines][clang-tidy-guidelines].
- **Check spelling**: NOT required to pass before the merge. It is up to the reviewer(s). See [here][spell-check-dict] if you want to add some words to the spell check dictionary.

[commit-guidelines]: https://www.conventionalcommits.org/en/v1.0.0/
[spell-check-dict]: https://github.com/tier4/autoware-spell-check-dict#how-to-contribute

### CODE REVIEW

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp [hunk 1]
```diff
@@ -228,241 +224,214 @@ void calcSlowDownPointsForPossibleCollision(
...
+  const double baselink_to_front = param.vehicle_info.baselink_to_front;
+  auto ll = path_lanelet.centerline2d();
+  for (const auto & dyn : dyn_objects) {
+    ArcCoordinates arc_coord_occlusion = getOcclusionPoint(dyn, ll);
+    ArcCoordinates arc_coord_occlusion_with_offset = {
+      arc_coord_occlusion.length - baselink_to_front,
+      calcSgnValue(arc_coord_occlusion.distance, half_vehicle_width)};
+    // ignore if collision is not avoidable by velocity control.
     if (
-      dyn.classification.at(0).label !=
```
Comment (taikitanaka3) [inline planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp:336]: filterParkedVehicle function remove unwanted vehicles instead

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp [hunk 2]
```diff
@@ -497,29 +465,38 @@ void generateSidewalkPossibleCollisions(
...
-    bool collision_free =
-      grid_utils::isCollisionFree(grid, occlusion_spot_position, intersection_point);
-    if (collision_free) {
-      PossibleCollisionInfo pc;
-      calculateCollisionPathPointFromOcclusionSpot(
-        pc, obstacle_point, offset_form_ego_to_target, path_lanelet, param);
-      if (pc.arc_lane_dist_at_collision.length < param.vehicle_info.baselink_to_front) {
-        continue;
-      }
-      possible_collisions.emplace_back(pc);
```
Comment (taikitanaka3) [inline planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp:519]: use nearest occlusion to path at each slice not "occlusion found at first".

**Comment by taikitanaka3** [comment]
image
image

**Comment by taikitanaka3** [comment]
tested by scenario sim
https://user-images.githubusercontent.com/65527974/151734912-c1db6488-f569-459e-a041-f1feb15cf9ac.mp4

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/risk_predictive_braking.cpp [hunk 3]
```diff
@@ -37,7 +37,7 @@ void applySafeVelocityConsideringPossibleCollison(
   for (auto & possible_collision : possible_collisions) {
     const double dist_to_collision = possible_collision.arc_lane_dist_at_collision.length;
     const double original_vel = possible_collision.collision_path_point.longitudinal_velocity_mps;
-    const double d_obs = possible_collision.arc_lane_dist_at_collision.distance;
+    const double d_obs = std::abs(possible_collision.arc_lane_dist_at_collision.distance);
```
Comment (taikitanaka3) [inline planning/behavior_velocity_planner/src/scene_module/occlusion_spot/risk_predictive_braking.cpp:40]: fixed implicitly ignored right area

File: planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp [hunk 4]
```diff
@@ -118,53 +142,98 @@ struct PossibleCollisionInfo
...
+  }
+}
+
+inline double offsetFromStartToEgo(
+  const PathWithLaneId & path, const Pose & ego_pose, const int closest_idx)
+{
+  double offset_from_ego_to_closest = 0;
+  for (int i = 0; i < closest_idx; i++) {
+    const auto & curr_p = path.points[i].point.pose.position;
+    const auto & next_p = path.points[i + 1].point.pose.position;
```
Comment (tkimura4) [inline planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp:163]: Is is possible that closest_idx will be the final point of the path.points?

Comment (tkimura4) [reply planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp:163]: Considering the unity with other parts, `.at(i)/.at(i+1)` is better here

File: planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp [hunk 5]
```diff
@@ -118,53 +142,98 @@ struct PossibleCollisionInfo
...
+  }
+}
+
+inline bool isStuckVehicle(PredictedObject obj, const double min_vel)
+{
+  if (
+    obj.classification.at(0).label == ObjectClassification::CAR ||
+    obj.classification.at(0).label == ObjectClassification::TRUCK ||
+    obj.classification.at(0).label == ObjectClassification::BUS) {
+    if (obj.kinematics.initial_twist_with_covariance.twist.linear.x < min_vel) {
```
Comment (tkimura4) [inline planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp:189]: std::abs?

File: planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp (hunk 4 above)
Comment (taikitanaka3) [reply planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp:163]: this is to find out ego closest point so if it's final point of path then ego goals

File: planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp (hunk 5 above)
Comment (taikitanaka3) [reply planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp:189]: maybe it's better for moving vehicle backward.

Comment (taikitanaka3) [inline planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp:189]:
```suggestion
    if (std::abs(obj.kinematics.initial_twist_with_covariance.twist.linear.x) < min_vel) {
```

Comment (taikitanaka3) [reply planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp]: fixed at 0f70107

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp [hunk 6]
```diff
@@ -202,19 +200,17 @@ void calcSlowDownPointsForPossibleCollision(
...
+        const double v0 = p_prev.longitudinal_velocity_mps;
+        const double v1 = p_next.longitudinal_velocity_mps;
+        const double v = getInterpolatedValue(d0, v0, dist_to_col, d1, v1);
         const double z = getInterpolatedValue(d0, p0.z, dist_to_col, d1, p1.z);
         // height is used to visualize marker correctly
-
-        possible_collisions[collision_index].collision_path_point.pose.position = setPoint(x, y, z);
-        possible_collisions[collision_index].intersection_pose.position.z = z;
-        possible_collisions[collision_index].obstacle_info.position.z = z;
+        auto & col = possible_collisions[collision_index];
```
Comment (tkimura4) [inline planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp:208]: .at()?

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp [hunk 7]
```diff
@@ -228,244 +224,210 @@ void calcSlowDownPointsForPossibleCollision(
...
+  arcs.pop_front();
+  // sort by arc distance
+  std::sort(arcs.begin(), arcs.end(), [](ArcCoordinates arc1, ArcCoordinates arc2) {
+    return std::abs(arc1.distance) < std::abs(arc2.distance);
+  });
+  // return closest to path point which is choosen by farthest 2 points.
+  return arcs.at(0);
+}
+
+double calcSgnValue(const double val, const double offset)
```
Comment (tkimura4) [inline planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp:278]: Can this function have a more descriptive name?

File: planning/behavior_velocity_planner/occlusion-spot-design.md [hunk 8]
```diff
@@ -91,31 +91,62 @@ Note that the accuracy and performance of this search method is limited due to t
...
+end note
+:handle collision offset;
+note right
+consider offset from path start to ego vehicle for possible collision
+end note
+:apply safe velocity consider possible collision;
+note right
+calculated by
+- ebs deceleration [m/s] emergency braking system consider lateral distance to the occlusion spot.
+- maximum allowed deceleration [m/s]
```
Comment (tkimura4) [inline planning/behavior_velocity_planner/occlusion-spot-design.md:144]: Is `[m/s]` a mistake of `[m/ss]`

Comment (taikitanaka3) [reply planning/behavior_velocity_planner/occlusion-spot-design.md:144]:
```suggestion
- maximum allowed deceleration [m/s^2]
```

File: planning/behavior_velocity_planner/occlusion-spot-design.md [hunk 9]
```diff
@@ -208,51 +228,22 @@ note right
...
-  - collision path point is calculated by arc coordinate consider ego vehicle's geometry.
-end note
 }
-:extract target road type start/end distance by arc length;
-:calculate original velocity and height for the possible collision;
-note right
-calculated (x,y,z,v) at the path using linear interpolation between interpolated path points.
-end note
-partition calculate_safe_velocity {
+partition handle_possible_collision{
```
Comment (tkimura4) [inline planning/behavior_velocity_planner/occlusion-spot-design.md:239]: Space is required before `{`

image

Comment (taikitanaka3) [reply planning/behavior_velocity_planner/occlusion-spot-design.md:239]:
```suggestion
partition handle_possible_collision {
```

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp (hunk 7 above)
Comment (taikitanaka3) [reply planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp:278]: calculateOffsetValue is good?

File: planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp (hunk 4 above)
Comment (taikitanaka3) [reply planning/behavior_velocity_planner/include/scene_module/occlusion_spot/occlusion_spot_utils.hpp]: fixed at c0d409a

File: planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp (hunk 6 above)
Comment (taikitanaka3) [reply planning/behavior_velocity_planner/src/scene_module/occlusion_spot/occlusion_spot_utils.cpp]: fixed at c0d409a

**Comment by tkimura4** [comment]
@taikitanaka3 
image

Only one virtual wall is visualized in Rviz.

**Comment by taikitanaka3** [comment]
@tkimura4 
since many wall may appear I limit it
https://tier4.atlassian.net/wiki/spaces/~484117074/pages/2268988030/occlusion+spot#VIrtual-Wall

**Comment by taikitanaka3** [comment]
@tkimura4 
fixed a drawing problem and change to acceptable slice params
image

**Review by tkimura4 — APPROVED** [review]
LGTM

(Please check this. @yukkysaito )

**Review by yukkysaito — APPROVED** [review]

**Pull request merged**